### ISSUE TITLE
Neuer HTTP Status 410 - Gone

### ISSUE THREAD

**Issue by Total-Reality** [contao/contao#3999]
### Description

Hi,
es wäre super, wenn ihr in der Seitenstruktur noch einen neuen Seitentyp vom Status "410" einführen könntet.

Es müsste also so ähnlich sein wie das Script:
https://github.com/contao/core-bundle/blob/5.x/src/Resources/contao/pages/PageError404.php

Es ist wichtig für Unterseiten, die dauerhaft nicht mehr existieren werden. Das kann z.B. nützlich sein, wenn es in einem Shop bestimmte Produkte dauerhaft nicht mehr geben wird. Oder Seiten bzw. Links, die ungewollt durch fehlerhafte Konfiguration in einer Erweiterung oder in Templates entstanden sind.
Google wird dann dazu gebracht, dass die Seiten sehr viel schneller aus dem Index herausfliegen. Google lernt dadurch, dass diese Seiten ungewollt sind und konzentriert seine Crawling-Power auf die anderen Seiten.
Im Prinzip ist 404 und 410 fast dasselbe. Allerdings ist ein 404 Fehler eher ein vorübergehender Zustand, während ein 410 Fehler langanhaltend bzw. für immer ist.

https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_client_errors
"Indicates that the resource requested is no longer available and will not be available again. This should be used when a resource has been intentionally removed and the resource should be purged. Upon receiving a 410 status code, the client should not request the resource in the future. Clients such as search engines should remove the resource from their indices.[41] Most use cases do not require clients and search engines to purge the resource, and a "404 Not Found" may be used instead."

Siehe auch hier:
https://www.sistrix.de/frag-sistrix/technisches-seo/status-codes/410/

Momentan ist es sehr umständlich das via htaccess umzusetzen:

1. Ich stelle global ein welche Seite im Fehlerfall angezeigt wird
ErrorDocument 410 /410-Page.html

Das ist dann allerdings eine Seite, die manuell gestylt und gepflegt wird und die leider nicht direkt aus Contao kommt.

2. Ich muss einzelne Seiten in der htaccess als 410 deklarieren

**Comment by aschempp** [comment]
Wie sollte die Seite denn generiert werden? Contao generiert ja die 404 wenn eine Seite nicht gefunden wird. Eine 410 würde nie generiert...

**Comment by fritzmg** [comment]
We could add general support for rendering a 410 page (by throwing the appropriate exception, just like 401, 403 or 404). How URLs are managed that should respond with 410 is another question and extensions could take advantage of that.

However, something like that could be implemented as an extension in any case I suppose.

**Comment by Total-Reality** [comment]
> We could add the general support for rendering a 410 page (by throwing the appropriate exception, just like 401, 403 or 404). How URLs are managed that should respond with 410 is another question and extensions could take advantage of that.

Yes, that could be the way to go.

> Wie sollte die Seite denn generiert werden? Contao generiert ja die 404 wenn eine Seite nicht gefunden wird. Eine 410 würde nie generiert...

Das stimmt. Ist derzeit so. 

Es gäbe aber folgende Fälle in denen ein 410 Status generiert werden könnte: 
- Extensions wie Shops oder Kataloge könnten die Verwaltung von Produkten ermöglichen, die es mal gab, aber nie wieder geben wird
- News im Backend auf unveröffentlicht stellen und danach festlegen, dass diese den Status 410 oder 404 bekommen
- Die gleiche Situation könnte es bei Seiten geben, also auf tl_page Ebene
- Festlegung einzelner Seiten als 410 Status via htaccess - u.a. damit die Seite auch schneller aus dem Google Index raus fliegt

Ich hab es z.B. schon häufiger erlebt, dass News aus folgenden Gründen unveröffentlicht worden sind: Rechtliche Gründe oder veraltete Informationen oder weil die News älter z.B. 4 Jahre sind, will man die nicht mehr darstellen. Das sind alles Gründe, dass die News dauerhaft raus ist, also 410.
Der tolle Vorteil ist nicht nur hinsichtlich der Suchmaschine. Man kann auch dem User dadurch genauere Infos liefern.
Zum Beispiel weiß er/sie/es dann, dass es bei 410 keinen Sinn ergibt weiter auf der Seite nach den Informationen zu suchen. Es gibt sie schlicht und ergreifend ganz **bewusst** nicht mehr.

Theoretisch gesehen könnte man auch noch Status 451 einführen: https://www.heise.de/newsticker/meldung/Neuer-Status-Code-451-zeigt-Zensur-an-3052138.html

Aber das wäre kein alles kein 'Must Have'. Entscheidend ist nur, dass man wie @fritzmg schon geschrieben hat den Status in der Seite festlegen kann, damit man dann via htaccess oder Extension diesen Status bespielen könnte.

**Comment by bennyborn** [comment]
+1 for implementing 410 pages

We actually needed that ourself for the relaunch of our own site. There was a good amount of old backlinks roaming around the web where we wanted to make sure that these are gone for good. We ended up implementing a custom route provider but having this in the core would be nice.

**Comment by m-vo** [comment]
That sounds more like a feature for [`terminal42/contao-url-rewrite`](https://github.com/terminal42/contao-url-rewrite) to me.

**Comment by fritzmg** [comment]
> That sounds more like a feature for [`terminal42/contao-url-rewrite`](https://github.com/terminal42/contao-url-rewrite) to me.

The implementation of the 410 page should be separate imho as the support to generate a 410 page is independent from actually triggering it. It could be a `vendor/contao-410-page` dependency that `terminal42/contao-url-rewrite` would require.

**Comment by rabauss** [comment]
This was also requested in an issue https://github.com/terminal42/contao-url-rewrite/issues/17.
Currently you can already provide a 410 as a response with `terminal42/contao-url-rewrite`, but this does not have a nice output like e.g. 404 pages

**Comment by aschempp** [comment]
I have a draft for Contao locally to support 410 and any other error page. But I think this is mostly confusing since you cannot trigger a 410. I would prefer to come up with a solution how to trigger 410 _before_ adding support to the core. And relying on extensions does not make much sense to me. If that were the case, the extension could simply render any page from the tree in that case.

**Comment by bennyborn** [comment]
> I would prefer to come up with a solution how to trigger 410 _before_ adding support to the core.

We mostly setup 410's once we relaunched a site. IMHO defining them in the `.htacces` should be a viable option

```apache
Redirect 410 /this/is/gone.html
````

**Comment by Total-Reality** [comment]
> I have a draft for Contao locally to support 410 and any other error page. But I think this is mostly confusing since you cannot trigger a 410. I would prefer to come up with a solution how to trigger 410 _before_ adding support to the core. And relying on extensions does not make much sense to me. If that were the case, the extension could simply render any page from the tree in that case.

I have described some use cases here:
https://github.com/contao/contao/issues/3999#issuecomment-1021298397

**Comment by aschempp** [comment]
Sorry I didn't mean use cases but a way for this to be used in plain Contao. If only and extension can generate a 410 error, then that extension can easily provide an option to generate such an error page (e.g. like any module by specifying a regular Contao page to be displayed).

**Comment by Total-Reality** [comment]
"Plain" Contao? You mean the Contao core? News and pages are in the core.

"then that extension can easily provide an option to generate such an error page"

It don't know if this is possible. I tried it out and it didn't work. I created a new page type error_410
https://www.fast-media.net/de/error-410-seite-entfernt/

However I was not able to declare one single page via htaccess as 410.

It still shows a white page with a big black "Gone".

It works only if I declare this in the htaccess:
`ErrorDocument 410 /410-Page.html`

But this page is static html. The contao page will be displayed never. Perhaps I forgot sth.?

**Comment by aschempp** [comment]
> It still shows a white page with a big black "Gone".

yes, because the extension displays that. That's what I mean. Contao has no knowledge of a 410 error, therefore it can never generate a 410 error page. Any controller can send a `410` HTTP status, that does not mean Contao will render such an error page.

**Comment by Total-Reality** [comment]
> yes, because the extension displays that.

Which extension? A white page is rendered by the hoster... 

> That's what I mean. Contao has no knowledge of a 410 error, therefore it can never generate a 410 error page.

I don't understand this. I created a `PageError410 ` class which extends the Frontend, but I think I forgot sth. to display it...


To test the general behavior I changed this line:
https://github.com/contao/contao/blob/5.x/core-bundle/src/Resources/contao/pages/PageError404.php#L48

to:
`header('HTTP/1.1 410 Gone');`

If I open a 404 page now, it works... the "redirect path" extension in Chrome shows: **410**

**Comment by aschempp** [comment]
> Which extension? A white page is rendered by the hoster...

How do you expect Contao to render a 410 page when your hoster sends a 410 status code? Have you ever tried to send a 404 from your .htaccess and have Contao render the error page?

**Comment by Total-Reality** [comment]
> How do you expect Contao to render a 410 page when your hoster sends a 410 status code? Have you ever tried to send a 404 from your .htaccess and have Contao render the error page?

Ah, okay. I understand the problem now. You are right, the 404 is rendered by the hoster then.
So we have no possibilty to set the 410 status via htaccess.

**Comment by bennyborn** [comment]
> So we have no possibilty to set the 410 status via htaccess.

I have to disagree with @aschempp. No matter if locally or on a hoster, all-inkl in this case, adding a 404 redirect in htaccess will trigger Contao's 404 page:

```htaccess
Redirect 404 /not-found
```
But to be fair this does not seem to be the case for 410's. In order for them to work we had to define an ErrorDocument:

```htaccess
ErrorDocument 410 /index.php
Redirect 410 /this-is-gone
```

**Comment by Total-Reality** [comment]
> I have to disagree with @aschempp. No matter if locally or on a hoster, all-inkl in this case, adding a 404 redirect in htaccess will trigger Contao's 404 page:
> 
> ```
> Redirect 404 /not-found
> ```

If I put this line in the htaccess, it will result in a 404 page rendered by the hoster (webgo):
`Redirect 404 /de/not-found/`

https://www.fast-media.net/de/not-found/

Every other page which doesn't exist results in a 404 rendered by Contao:
https://www.fast-media.net/de/not-found2/

Perhaps I made sth. wrong?

**Comment by fritzmg** [comment]
I think you guys are misusing these Apache directives? The Apache directive

```
Redirect 404 /not-found
```

means that Apache will respond with a 404 status code, if the URL `/not-found` is accessed. Using it the way you guys are doing does not really make sense.

```
ErrorDocument 404 /index.html
```

means that when a non-existing resource is accessed in a request, Apache will rewrite the URL internally to the given path. i.e. the contents of `/index.html` will be in the response. See https://httpd.apache.org/docs/2.4/mod/core.html#errordocument

All of this is Apache specific and has nothing to do with Contao itself and neither does it have to do anything with the proposed 410 page type in Contao.

**Comment by bennyborn** [comment]
@fritzmg You're right. I totally overlooked the important line in our .htaccess

```
ErrorDocument 404 /seite-nicht-gefunden.html
```

> All of this is Apache specific and has nothing to do with Contao itself and neither does it have to do anything with the proposed 410 page type in Contao.

Well, at least it shows it is technically absolutely possible to force Contao to render a 404 or 410. The latter would of course only be possible if Contao provided such a pagetype. 

IMHO there should be no discussion about implementing such a type if there are already ways to trigger such pages albeit webserver specific 🤷‍♂️

**Comment by fritzmg** [comment]
> Well, at least it shows it is technically absolutely possible to force Contao to render a 404 or 410.

I do not know what you mean. The above directives are Apache specific and have nothing to do with Contao. Furthermore Contao supports the 404 page type natively, but not a 410 page type yet - which is what this issue is about.

Please keep the discussions about Apache directives separate and use an appropriate community for that. Otherwise it dilutes this issue with completely irrelevant comments.

**Comment by bennyborn** [comment]
Back to topic: I think there is an actual need for Contao to be able to generate a 410 page. If it isn't shipped with the Core it could be a simple `vendor/contao-410-page` package like suggested 👍

**Comment by fritzmg** [comment]
> There is actually not even a way to trigger a 404.

Of course there is? 404 is fully supported.

> It is only triggered by unpublishing existing pages / news articles etc. The Core itself currently has no way to trigger a 404 for a page that does not exist in the site structure at all.

That statement is false. Any URL that is processed by the application will show the 404 page, if it cannot be matched in the router.

**Comment by leofeyer** [comment]
As discussed in the Contao call, we could add a "gone" checkbox to regular pages and throw a `GoneException` if the page is rendered. Then a new error_410 page could be triggered.

**Comment by eki89** [comment]
I tried to implement a 410 error page as a Contao extension. However, this requires an adjustment to the PrettyErrorScreenListener: https://github.com/contao/contao/blob/5.x/core-bundle/src/EventListener/PrettyErrorScreenListener.php#L99

Do you think 410 error pages should become a core feature in the future? Or should a technique be implemented to be able to implement further error pages via thrid party extensions?

Personally, I tend to implement it as a core feature, because I don't see any other aditionall useful error pages. In this case the implementation should be very easy (at least for Contao 5.2).

**Comment by asaage** [comment]
> we could add a "gone" checkbox to regular pages and throw a GoneException if the page is rendered.

I think thats not a good approach as It requieres to keep the regular/original page and if you keept it its obviously not "gone".
https://github.com/terminal42/contao-url-rewrite has a way to send 410 by matching aliases.
wasn't this extension to be incorporated into a future contao version anyway?

**Comment by eki89** [comment]
I agree with you completely. Declaring individual pages as gone via a checkbox is inconsistent.

It seems to make sense for us to declare pages via a checkbox and throw a GoneException by throwing a PageNotFoundException (if the checkbox is activated for the current page). For example a reader module could throw a GoneException and display the corresponding error page when a PageNotFoundException is thrown. In case of news, 410 errors would always be thrown if the reader module is called and no matching news could be found. This solution could be combined with all reader modules of third-party extensions.

Of course, this solution isn't perfect as well...

**Comment by bennyborn** [comment]
If anyone is still looking for a "solution" we've just released an exension that adds a new page type for that https://github.com/numero2/contao-page-410
